pkg/blocker/blocker.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package blocker
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+type peer struct {
+	blockAfter time.Time // timestamp of the point we've timed-out or got an error from a peer
+	addr       swarm.Address
+}
+
+type Blocker struct {
+	mux           sync.Mutex
+	disconnector  p2p.Blocklister
+	flagTimeout   time.Duration // how long before blocking a flagged peer
+	blockDuration time.Duration // how long to blocklist a bad peer
+	workerWakeup  time.Duration // how long to blocklist a bad peer
+	peers         map[string]*peer
+	logger        logging.Logger
+	wakeupCh      chan struct{}
+	quit          chan struct{}
+}
+
+func New(dis p2p.Blocklister, flagTimeout, blockDuration, wakeUpTime time.Duration, logger logging.Logger) *Blocker {
+
+	b := &Blocker{
+		disconnector:  dis,
+		flagTimeout:   flagTimeout,
+		blockDuration: blockDuration,
+		workerWakeup:  wakeUpTime,
+		peers:         map[string]*peer{},
+		wakeupCh:      make(chan struct{}),
+		quit:          make(chan struct{}),
+		logger:        logger,
+	}
+
+	go b.run()
+
+	return b
+}
+
+func (b *Blocker) run() {
+
+	for {
+		select {
+		case <-b.quit:
+			return
+		case <-time.After(b.workerWakeup):
+			b.block()
+		}
+	}
+}
+
+func (b *Blocker) block() {
+	b.mux.Lock()
+	defer b.mux.Unlock()
+
+	for key, peer := range b.peers {
+		if !peer.blockAfter.IsZero() && time.Now().After(peer.blockAfter) {
+			if err := b.disconnector.Blocklist(peer.addr, b.blockDuration, "blocker: flag timeout"); err != nil {
```

Comment:
would be good to double check this doesn't panic on shutdown because of libp2p shutdown